sample_consensus/include/pcl/sample_consensus/impl/sac_model_torus.hpp

```diff
@@ -0,0 +1,518 @@
+void
+pcl::SampleConsensusModelTorus<PointT, PointNT>::optimizeModelCoefficients(
+    const Indices& inliers,
+    const Eigen::VectorXf& model_coefficients,
+    Eigen::VectorXf& optimized_coefficients) const
+{
+
+  optimized_coefficients = model_coefficients;
+
+  // Needs a set of valid model coefficients
+  if (!isModelValid(model_coefficients)) {
+    PCL_ERROR("[pcl::SampleConsensusModelTorus::optimizeModelCoefficients] Given model "
+              "is invalid!\n");
+    return;
+  }
+
+  // Need more than the minimum sample size to make a difference
+  if (inliers.size() <= sample_size_) {
+    PCL_ERROR("[pcl::SampleConsensusModelTorus::optimizeModelCoefficients] Not enough "
+              "inliers to refine/optimize the model's coefficients (%lu)! Returning "
+              "the same coefficients.\n",
+              inliers.size());
+    return;
+  }
+
+  OptimizationFunctor functor(this, inliers);
+  Eigen::NumericalDiff<OptimizationFunctor> num_diff(functor);
+  Eigen::LevenbergMarquardt<Eigen::NumericalDiff<OptimizationFunctor>, double> lm(
+      num_diff);
+  Eigen::VectorXd coeff(model_size_);
+  int info = lm.minimize(coeff);
+
+  if (info) {
+    PCL_ERROR("[pcl::SampleConsensusModelTorus::optimizeModelCoefficients] Not enough "
+              "inliers to refine/optimize the model's coefficients (%lu)! Returning "
+              "the same coefficients.\n",
+              inliers.size());
+    return;
+  }
+  else {
+    for (Eigen::Index i = 0; i < coeff.size(); ++i)
+      optimized_coefficients[i] = static_cast<float>(coeff[i]);
+  }
+}
+
+//////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+template <typename PointT, typename PointNT>
+void
+pcl::SampleConsensusModelTorus<PointT, PointNT>::projectPointToTorus(
+    const Eigen::Vector3f& p_in,
+    const Eigen::VectorXf& model_coefficients,
+    Eigen::Vector3f& pt_out) const
+{
+
+  // Fetch optimization parameters
+  const float& R = model_coefficients[0];
+  const float& r = model_coefficients[1];
+
+  const float& x0 = model_coefficients[2];
+  const float& y0 = model_coefficients[3];
+  const float& z0 = model_coefficients[4];
+
+  const float& nx = model_coefficients[5];
+  const float& ny = model_coefficients[6];
+  const float& nz = model_coefficients[7];
+
+  // Normal of the plane where the torus circle lies
+  Eigen::Vector3f n{nx, ny, nz};
+  n.normalize();
+
+  Eigen::Vector3f pt0{x0, y0, z0};
+
+  // Ax + By + Cz + D = 0
+  float D = -n.dot(pt0);
+  Eigen::Vector4f planeCoeffs{n[0], n[1], n[2], D};
+  planeCoeffs.normalized();
+  Eigen::Vector3f p(p_in);
+
+  // Project to the torus circle plane
+  Eigen::Vector3f pt_proj;
+  projectPointToPlane(p, planeCoeffs, pt_proj);
+
+  // TODO expect singularities, mainly pt_proj_e == pt0 || pt_e ==
+  // Closest point from the inner circle to the current point
+  Eigen::Vector3f circle_closest;
+  circle_closest = (pt_proj - pt0).normalized() * R + pt0;
```

Comment:
I think this can fail for a "self-intersecting spindle torus" (R < r?) for points in the middle

Reply:
You are right, also this solution doesn't really leverage the use of the normals either. Let me come back with a solution that is more inspired.